File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.java
package io.appium.espressoserver.lib.handlers;

import android.app.Activity;

import java.util.List;

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException;
import io.appium.espressoserver.lib.helpers.ActivityHelper;
import io.appium.espressoserver.lib.helpers.BackdoorUtils;
import io.appium.espressoserver.lib.helpers.InvocationOperation;
import io.appium.espressoserver.lib.model.MobileBackdoorParams;

import static io.appium.espressoserver.lib.helpers.AndroidLogger.logger;

public class MobileBackdoor implements RequestHandler<MobileBackdoorParams, Object> {

    @Override
    public Object handle(final MobileBackdoorParams params) throws AppiumException {
        logger.info("Invoking Backdoor");
        if (params.getTarget() == null) {
            throw new InvalidArgumentException("Target must not be empty and must be of type: 'activity', 'application'");
        }

        Activity activity = ActivityHelper.getCurrentActivity();
        List<InvocationOperation> ops = BackdoorUtils.getOperations(params.getMethods());

        switch (params.getTarget()) {
            case ACTIVITY:
                return BackdoorUtils.invokeMethods(activity, ops);
            case APPLICATION:
                return BackdoorUtils.invokeMethods(activity.getApplication(), ops);
            default:
                throw new InvalidArgumentException(String.format("target cannot be %s", params.getTarget()));
        }

    }

}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt
package io.appium.espressoserver.lib.handlers

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
import io.appium.espressoserver.lib.helpers.ActivityHelpers
import io.appium.espressoserver.lib.helpers.InvocationOperation
import io.appium.espressoserver.lib.helpers.reflection.ClassUtils
import io.appium.espressoserver.lib.model.EspressoElement
import io.appium.espressoserver.lib.model.InvocationTarget
import io.appium.espressoserver.lib.model.MobileBackdoorParams

class MobileBackdoor : RequestHandler<MobileBackdoorParams, Any?> {

    @Throws(AppiumException::class)
    override fun handleInternal(params: MobileBackdoorParams): Any? {
        params.target?.let {target ->
            val activity = ActivityHelpers.currentActivity
            val ops = getBackdoorOperations(params)

            @Suppress("REDUNDANT_ELSE_IN_WHEN")
            val result = when (target) {
                InvocationTarget.ACTIVITY -> invokeBackdoorMethods(activity, ops)
                InvocationTarget.APPLICATION -> invokeBackdoorMethods(activity.application, ops)
                InvocationTarget.ELEMENT -> invokeBackdoorMethods(EspressoElement.getViewById(params.targetElement), ops)
                else -> throw InvalidArgumentException("target cannot be '$target'")
            } ?: return null

            if (result is Array<*> && result.filterNotNull()
                    .all { it is String || (ClassUtils.wrapperToPrimitive(it.javaClass) != null) }
            ) {
                return result
            }

            ClassUtils.wrapperToPrimitive(result.javaClass)?.let { return result }
                ?: return result.toString()
        }

        throw InvalidArgumentException("Target must not be empty and must be of type: 'activity', 'application'")
    }

    @Throws(AppiumException::class)
    private fun invokeBackdoorMethods(invokeOn: Any, ops: List<InvocationOperation>): Any? {
        return ops.fold(invokeOn) { invocationTarget, operation -> operation.apply(invocationTarget) }
    }

    @Throws(InvalidArgumentException::class)
    private fun getBackdoorOperations(params: MobileBackdoorParams): List<InvocationOperation> {
        return params.methods.map {method ->
            InvocationOperation(method.name, method.arguments, method.argumentTypes)
        }
    }
}
